onadata/apps/api/tests/viewsets/test_abstract_viewset.py

```diff
@@ -693,9 +699,82 @@ def _publish_markdown(self, md, user, project=None, **kwargs):
+        |          | integer            | circumference                              | Tree circumference in cm | circumference_cm                           |
+        |          | text               | intake_notes                               | Intake notes             |                                            |
+        | choices  |                    |                                            |                          |                                            |
+        |          | list_name          | name                                       | label                    |                                            |
+        |          | species            | wallaba                                    | Wallaba                  |                                            |
+        |          | species            | mora                                       | Mora                     |                                            |
+        |          | species            | purpleheart                                | Purpleheart              |                                            |
+        |          | species            | greenheart                                 | Greenheart               |                                            |
+        | settings |                    |                                            |                          |                                            |
+        |          | form_title         | form_id                                    | version                  | instance_name                              |
+        |          | Trees registration | trees_registration                         | 2022110901               | concat(${circumference}, "cm ", ${species})|
+        | entities |                    |                                            |                          |                                            |
+        |          | list_name          | label                                      |                          |                                            |
+        |          | trees              | concat(${circumference}, "cm ", ${species})|                          |                                            |"""
+
+        if not hasattr(self, "project"):
+            self._project_create()
+        elif self.project.created_by != self.user:
+            self._project_create()
+
+        data_dict = self._publish_markdown(
+            md,
+            self.user,
+            self.project,
+            id_string="trees_registration",
+            title="Trees registration",
+        )
+        latest_form = XForm.objects.all().order_by("-pk").first()
+        XFormVersion.objects.create(
+            xform=latest_form,
+            version="2022110901",
+            xml=data_dict.xml,
+            json=json.dumps(data_dict.json),
+        )
```

Comment:
Could we have used `onadata.libs.utils.logger_tools.create_xform_version(xform: Xform, user: User)`?


Perhaps, could we consider adding `publish_md_form()` similar to the `publish_xls_form()` and `publish_xml_form()` functions under `onadata.libs.utils.logger_tools` that have the same signature and all create the `XFormVersion` object of the form they publish? This will mean we update `self._publish_markdown()` to use the new `publish_md/markdown_form()` function and an `XFormVersion` object will always be created when publishing the form.

Reply:
@kelvin-muchiri There are several places where the `XFormVersion` is directly created. So, if we apply this change only in `self._publish_markdown()`, the repetition can be avoided.

Reply:
@ukanga I resolved this by [removing the redundant ](https://github.com/onaio/onadata/pull/2504/files#diff-d5e812684c4b5ff59d4809a88e9ad052819979735ef01935a7ad1688a0aab6dfR101) `_publish_markdown` in `TestAbstractViewSet`